specs/observability.md

```diff
@@ -0,0 +1,339 @@
+
+---
+
+## Metrics Collection
+
+### Analytics Engine Integration
+
+The system MUST integrate with Cloudflare Analytics Engine for metrics storage and querying. Metrics MUST be recorded with:
+
+- **User identifier:** Anonymized (SHA-256 hash of user ID)
+- **Metric name:** Descriptive identifier (e.g., `tool_read_success`, `storage_quota_bytes`)
+- **Metric value:** Numeric value (duration, count, bytes, etc.)
+- **Timestamp:** Automatic timestamp by Analytics Engine
+
+### Metric Recording Points
+
+Metrics MUST be recorded at the following points:
+
+**Tool execution:**
+- Record start time
+- Record success/failure outcome
+- Record duration
+- Record tool-specific metadata (file size for reads, etc.)
+
+**Rate limiting:**
+- Record rate limit checks
+- Record rate limit violations
+- Record limit type (minute, hour, day)
+
+**Storage operations:**
+- Record current usage on quota checks
+- Record operation type (read, write, delete)
+- Record operation success/failure
+
+**OAuth events:**
+- Record authentication attempts
+- Record success/failure outcomes
+- Record token refresh events
+
+### Silent Failure Requirement
+
+Metrics collection failures MUST NOT impact application functionality. If Analytics Engine is unavailable or metric recording fails, the system MUST:
+- Continue serving user requests normally
+- Log the telemetry failure at DEBUG or WARN level
+- Not retry metric recording (avoid cascading failures)
+- Not surface errors to users
+
+---
+
+## Alerting Requirements
+
+### Alert Configuration
+
+The system SHOULD support alerting on the following conditions:
+
+**Critical alerts (immediate action required):**
+- Error rate >5% sustained for 5 minutes
+- Authentication failure spike (>5 failed attempts in 5 minutes from same user)
+- Storage quota exceeded (>10GB)
+- Any path traversal attempt detected
+
+**Warning alerts (investigate within 24 hours):**
+- Request spike (>10x average sustained for 5 minutes)
+- Storage approaching quota (>8GB, 80% of limit)
+- Rate limit hit rate >10%
+- Backup failure
+- Performance degradation (p95 >1000ms sustained)
+
+**Informational alerts (review weekly):**
+- Cost metrics approaching thresholds
+- Usage pattern changes
+- New error types appearing
+
+### Alert Delivery
+
+**[DEFERRED]** Alert delivery mechanisms (email, webhook, PagerDuty, etc.) are implementation decisions. The requirement is that alerts can be configured and delivered reliably.
+
+---
+
+## Dashboard Requirements
+
+### Cloudflare Dashboard
+
+The system MUST leverage the built-in Cloudflare Workers dashboard for:
+- Real-time request volume and error rate
+- CPU time and duration percentiles
+- Request volume by status code
+- Geographic distribution
+
+### Custom Analytics
+
+The system SHOULD support custom dashboards querying Analytics Engine data for:
```

Comment:
Accessing and visualising this data is out of scope right now. All that's important is that the data is being collected and stored so that we can work with it later but we'll decide what we any to do another time.